spring-integration-mail/src/main/java/org/springframework/integration/mail/AbstractMailReceiver.java

```diff
@@ -298,28 +328,64 @@ private Folder obtainFolderInstance() throws MessagingException {
 					this.logger.debug("Received " + messages.length + " messages");
 				}
 
-				Message[] filteredMessages = filterMessagesThruSelector(messages);
+				MimeMessage[] filteredMessages = filterMessagesThruSelector(messages);
 
 				postProcessFilteredMessages(filteredMessages);
 
-				return filteredMessages;
+				if (this.headerMapper != null) {
+					org.springframework.messaging.Message<?>[] converted =
+							new org.springframework.messaging.Message<?>[filteredMessages.length];
+					int n = 0;
+					for (MimeMessage message : filteredMessages) {
+						converted[n++] = getMessageBuilderFactory().withPayload(extractContent(message))
+								.copyHeaders(this.headerMapper.toHeaders(message))
+								.build();
+					}
+					return converted;
+				}
+				else {
+					return filteredMessages;
+				}
 			}
 			finally {
 				MailTransportUtils.closeFolder(this.folder, this.shouldDeleteMessages);
 			}
 		}
 	}
 
+	private Object extractContent(MimeMessage message) {
+		Object content;
+		try {
+			content = message.getContent();
+			if (content instanceof InputStream) {
+				ByteArrayOutputStream baos = new ByteArrayOutputStream();
+				FileCopyUtils.copy((InputStream) content, baos);
+				content = baos.toByteArray();
+			}
+			else if (content instanceof Multipart && this.multipartAsBytes) {
+				ByteArrayOutputStream baos = new ByteArrayOutputStream();
+				((Multipart) content).writeTo(baos);
+				content = baos.toByteArray();
+			}
+			return content;
```

Comment:
I think for common `content` object we should populate the `Content-Type` header as well: http://stackoverflow.com/questions/11240368/how-to-read-text-inside-body-of-mail-using-javax-mail. To let end-user to decide what to do further with the result.
Although for those `baos.toByteArray()`, I'd populate `MessageHeaders.CONTENT_TYPE` as an `application/octet-stream`, too.
WDYT?